pkg/controller/che/oauth_initial_htpasswd_provider.go

```diff
@@ -0,0 +1,248 @@
+//
+// Copyright (c) 2012-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package che
+
+import (
+	"context"
+
+	errorMsg "errors"
+
+	"github.com/eclipse/che-operator/pkg/deploy"
+	"github.com/eclipse/che-operator/pkg/util"
+	configv1 "github.com/openshift/api/config/v1"
+	oauthv1 "github.com/openshift/api/config/v1"
+	userv1 "github.com/openshift/api/user/v1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	htpasswdIdentityProviderName        = "htpasswd-eclipse-che"
+	htpasswdSecretName                  = "htpasswd-eclipse-che"
+	ocConfigNamespace                   = "openshift-config"
+	openShiftOAuthUserCredentialsSecret = "openshift-oauth-user-credentials"
+)
+
+// OpenShiftOAuthUserHandler - handler to create or delete new Openshift oAuth user.
+type OpenShiftOAuthUserHandler interface {
+	CreateOAuthInitialUser(openshiftOAuth *oauthv1.OAuth, deployContext *deploy.DeployContext) error
+	DeleteOAuthInitialUser(deployContext *deploy.DeployContext) error
+}
+
+// OpenShiftOAuthUserOperatorHandler - implementation OpenShiftOAuthUserHandler.
+type OpenShiftOAuthUserOperatorHandler struct {
+	OpenShiftOAuthUserHandler
+	runtimeClient client.Client
+	runnable      util.Runnable
+}
+
+// NewOpenShiftOAuthUserHandler - create new OpenShiftOAuthUserHandler instance
+func NewOpenShiftOAuthUserHandler(runtimeClient client.Client) OpenShiftOAuthUserHandler {
+	return &OpenShiftOAuthUserOperatorHandler{
+		runtimeClient: runtimeClient,
+		// real process implementation. In the test we are using mock
+		runnable: util.NewRunnable(),
+	}
+}
+
+// CreateOAuthInitialUser - creates new htpasswd provider with inital user with Che flavor name
+// if Openshift cluster hasn't got identity providers, otherwise do nothing.
+// It usefull for good first user expirience.
+// User can't use kube:admin or system:admin user in the Openshift oAuth. That's why we provide
+// initial user for good first meeting with Eclipse Che.
+func (iuh *OpenShiftOAuthUserOperatorHandler) CreateOAuthInitialUser(openshiftOAuth *oauthv1.OAuth, deployContext *deploy.DeployContext) error {
```

Comment:
```suggestion
func (iuh *OpenShiftOAuthUserOperatorHandler) SyncOAuthInitialUser(openshiftOAuth *oauthv1.OAuth, deployContext *deploy.DeployContext) (bool, error) {
```

Reply:
1. Generate password once, set in variable
2. Generate `htpasswdFileContent` once, set in variable
3. Check if credentials secret exists != password, update password in variable and `htpasswdFileContent` in variable
4. Sync credentials secret
5. Sync OpenShift secret
6. Append identity provider